### PR TITLE
feat(#54): add Egypt National ID (EGY) validation and parsing

### LINKED ISSUE #54: feat(egy): Add Egypt national ID validator
## Description
Add support for Egypt National ID (الرقم القومي).

## Parent Epic
#14 - Add support for new countries

## Egypt National ID Format
- 14 digits
- Format: CYYMMDDSSRRRRN
- C: Century (2=1900s, 3=2000s)
- YYMMDD: Birth date
- SS: Governorate code
- RRRR: Sequence number
- N: Check digit

## Tasks
- [ ] Research and verify format
- [ ] Create `src/countries/egy/` directory
- [ ] Implement NationalId validator class
- [ ] Implement validate() with checksum
- [ ] Implement parse() function
- [ ] Add to main dispatcher
- [ ] Add to getCountryIdFormat()
- [ ] Write comprehensive tests (80%+ coverage)

## Acceptance Criteria
- [ ] Validation works correctly
- [ ] Parse extracts birth date, gender, governorate
- [ ] Test coverage 80%+
- [ ] Format documented

## Labels
enhancement, new-country, egypt

### PR DESCRIPTION
**Pull request by ramezhanna**
Closes #54.

Adds support for the **Egyptian National ID** (الرقم القومي), a 14-digit number `CYYMMDDGGSSSSV`, following the established country-module + registry pattern.

## What it does
- **`validate()`** — format + semantic validation: length, century (`2`→1900s, `3`→2000s), a **real calendar date** (rejects e.g. `2099-02-29`), and a **known governorate code**. Non-string input is rejected rather than coerced.
- **`parse()`** — extracts `birthDate`, `gender` (serial parity), `governorateCode`, `governorate` (name), `serialNumber`, `checksum`, `age`. The `checksum` field reports the ID's own trailing digit as-is; it is not verified.
- **`checksum()`** — intentionally unimplemented, always returns `null`, mirroring `hasChecksum: false` (the `NGA`/`NZL`/`VNM` precedent).
- Registered as **`EGY`** with **`EG`** alias; exported from the package root. `validateNationalId`, `parseIdInfo`, and `getCountryIdFormat` all work for Egypt automatically.

## Checksum decision (important)
Egypt publishes **no official check-digit algorithm**. Wikipedia calls digit 14 only a "check digit"; the "Luhn" claim comes from blogs with no verified vectors, and the Python upstream has no `egy` module.

A weighted mod-11 rule was investigated and **rejected as underdetermined**: it reproduces the check digit on all 5 IDs it was fitted to, but those IDs cannot identify it — all 5 are governorate `01` with a serial below 1000, so digits 8 and 10 are `0` in every sample and their weights are multiplied by zero. Roughly 10⁸ weight vectors satisfy the same corpus while disagreeing outside governorate `01`–`04`.

Following the repo's **Bahrain CPR precedent**, validation is therefore **format + semantic only** (`hasChecksum: false`), and no check digit is computed or enforced — so genuine IDs are never silently rejected. The hypothesis and the reasoning that rejected it are preserved in [`docs/research/egypt-national-id.md`](https://github.com/ramezhanna/idnumbers-npm/blob/feat/54-egypt-national-id/docs/research/egypt-national-id.md), not in dead code.

## Tests & verification
- New suite `issue-54-egy-nationalId.test.ts`: **42 cases**, **98% stmt / 89% branch / 100% func** coverage for the module.
- Full suite: **2193 passing** across 29 suites. `npm run build`, `npm run lint`, `npm run format:check` all pass.
- Bumped registry count 80 → 81 in the migration parity test and README.
- EGY cases live in the issue-54 suite rather than `country-specific-validation.test.ts`, which asserts parity with the Python `idnumbers` package — Python has no `egy` module to compare against.

> Note on a non-upstream country: Egypt was previously skipped as "not in Python idnumbers", but **RUS** was already added under the same condition — precedent for supporting countries absent from upstream.

> All example IDs are **synthetic**, used only to exercise validation.

### CODE REVIEW

**Comment by ramezhanna** [comment]
## Follow-up: cross-cutting parity test coverage (commit `2dceae3`)

Extended the migration / comprehensive suites so Egypt is exercised through the same per-country parity paths as every other country. **This also surfaced and fixed a real gap** in the first commit.

### 🐛 Gap found & fixed
EGY was registered in the registry but missing from `SUPPORTED_COUNTRIES` in `src/index.ts`. As a result:
- `listSupportedCountries()` **omitted Egypt**, and
- `getCountryIdFormat('EGY').countryName` returned `"EGY"` instead of `"Egypt"`.

Added `{ code: 'EGY', name: 'Egypt', idType: 'National ID' }`. Verified at runtime → `countryName: Egypt | idType: National ID | inList: true`.

### ✅ Test additions
| File | Change |
| --- | --- |
| `comprehensive-validation.test.ts` | Egypt validate + parse assertions in **African Countries**; un-skipped Egypt in the **Parse ID Information** block; fixed stale "80 countries" comment |
| `country-specific-validation.test.ts` | Replaced the stale `// EGY … removed` note with a real **EGY describe block** (3 valid + 4 invalid vectors) |
| `getCountryIdFormat-migration.test.ts` | EGY row in the country table + `EG` alias test; `80 → 81` docstring |
| `parseIdInfo-migration.test.ts` | EGY row in the per-country parse parity table (plus the count/key/alias bump from commit 1) |
| `validateNationalId-migration.test.ts` | EGY row in the per-country validation parity table |

### ⏸️ Deliberately unchanged
- **`utils.test.ts`** — `luhnDigit` is already covered generically (both branches); a country-specific case there would be misplaced. Egypt's checksum path is tested in `issue-54-egy-nationalId.test.ts`.
- **`registry.test.ts`** — uses fresh throwaway registries (USA/DEU/AUS mocks), no global country count, so nothing to update.

### Verification
`npm test` → **2159 passing** (was 2141, +18) · `npm run build`, `npm run lint`, `npm run format:check` all clean · pre-commit hook passed.

**Review by AngusHsu — CHANGES_REQUESTED** [review]
## Summary
Solid, well-researched addition (Egypt National ID, 14-digit format + weighted mod-11 checksum). All CI checks pass (Prettier, ESLint, `tsc`, Jest 2156/2156). Two concrete correctness/test-coverage gaps and one disclosed-but-consequential product risk need addressing before merge.

## Status: NEEDS CHANGES

## Scores
- Security: 10/10
- Performance: 10/10
- Code Quality: 7/10
- Test Coverage: 8/10

## Issues Found

### P0 - Critical (BLOCKED)
None.

### P1 - Important (NEEDS CHANGES)

1. **`src/countries/egy/nationalId.ts:204`** `parse()` doesn't verify the checksum, so it disagrees with `validate()` on the same input `[Codex]`
   - Issue: `parse()` returns `{ isValid: true, ... }` as soon as format + date + governorate check out — it never compares the trailing digit against `checksum()`. But `METADATA.hasChecksum = true` and `validate()` **does** enforce it.
   - Impact: `validateNationalId('EGY', '29001010100010')` → `isValid: false` (bad check digit), while `parseIdInfo('EGY', '29001010100010')` → truthy result with `isValid: true` and a full demographic breakdown, for the exact same string. Callers who call `parse`/`parseIdInfo` without also calling `validate` will silently accept IDs with wrong check digits.
   - Precedent: Kuwait (`src/countries/kwt/index.ts`), which has the same "checksum enforced" contract, verifies the checksum inside `parse()` and returns `null` on mismatch — `validate()` there is literally `parse(idNumber) !== null`. EGY should follow the same pattern.
   - Fix: in `parse()`, compute `checksum(normalized)` and return `null` when it doesn't match `parseInt(check, 10)`.

2. **`src/__tests__/getCountryIdFormat-migration.test.ts`** EGY's `displayFormat` is never asserted anywhere in the suite `[Codex]`
   - Issue: EGY was added to the "registered countries" list (idType check, line 60) and the alias list (line 163), but **not** to the separate `countriesWithFormat` list (lines 120-139) that asserts `getCountryIdFormat(code).format` equals the expected display-format string (the pattern KWT/KAZ/etc. already follow). `src/__tests__/issue-54-egy-nationalId.test.ts`'s own `getCountryIdFormat` test also stops short of checking `.format`.
   - Impact: nothing in CI actually pins `CYYMMDDGGSSSSV` as the returned format string — a future refactor could silently change/drop it with all tests still green.
   - Per CLAUDE.md: "`getCountryIdFormat-migration.test.ts` fixtures must be updated in lockstep with any `METADATA` format changes."
   - Fix: add `{ code: 'EGY', format: 'CYYMMDDGGSSSSV' }` to `countriesWithFormat`.

3. **`README.md:477`** stale test count `[Multi: Codex+Manual]`
   - Issue: says "comprehensive test coverage with 2105 tests"; a full `npx jest` run on this branch reports **2156 passed, 2156 total**.
   - Fix: bump 2105 → 2156, per the project's own convention (and PR #113's precedent).

4. **`src/countries/egy/nationalId.ts:78,127-165` + `README.md`** checksum enforced-by-default on n=5 unpublished-sample evidence, undisclosed to README readers `[Manual]`
   - Issue: `hasChecksum: true` makes `validate()` hard-reject any ID whose 14th digit doesn't match a weighted mod-11 formula that has **no official government citation** and **no Python-upstream parity baseline** — adopted solely because it matched 5 real (unpublished, unreproducible) IDs. This is the opposite choice the repo made for the structurally identical Bahrain CPR case (`hasChecksum: false`, precedent explicitly named in `docs/research/egypt-national-id.md`).
   - Concrete edge case: the `expected === 10` branch (line 127-128) hard-rejects that payload as "no valid ID can exist" — a claim the author's own doc flags as **unverified** (`docs/research/egypt-national-id.md`, "Call for contributions" section).
   - Disclosure gap: the caveat lives in `docs/research/` and the `METADATA.checksumAlgorithm` string — both fine for someone auditing the code, but the README's Africa section only adds the plain country line with zero caveat, so an npm consumer evaluating "can I trust `validate()`" has no visible warning.
   - This is a judgment call, not a defect — recommend either (a) a short README caveat that the check digit is empirically derived from a small sample, not an official spec, and/or (b) revisiting whether checksum enforcement should be opt-in (matching the Bahrain precedent) until an authoritative source confirms the algorithm.

### P2 - Suggestions (Optional)
1. **`src/countries/egy/nationalId.ts:170`** `parse()` docstring says "(NOT the opt-in checksum)" — the checksum is no longer opt-in (enforced by default via `hasChecksum: true`); only the adjective is stale, the underlying fact is correct. `[Manual]`
2. **`src/__tests__/getCountryIdFormat-migration.test.ts:11`** inline comment still says "All 80 registered countries" while the docblock above it already says 81. `[Manual]`
3. Two commits in this PR (`docs: update README to reflect support for 81 countries`, `docs: clarify checksum description for Egypt National ID`) omit the `(#54)` issue tag that the rest of the PR's commits use. `[Manual]`
4. FYI, out of this PR's scope: `src/countries/nzl/irdNumber.ts:9` has an "exactly-80 primary-key invariant" comment now stale due to the 80→81 bump — pre-existing file, not touched by this PR. `[Manual]`

## Review Sources
- Manual + 5-agent multi-perspective review (`reviewer` agent): checksum-enforcement risk assessment (P1 #4), all P2s, corroborated the stale test count
- Codex (gpt-5.5, high effort): parse()/checksum inconsistency (P1 #1), missing format-fixture (P1 #2), stale test count (P1 #3)
- Independently verified by me: reproduced all 4 P1s directly against source (confirmed `parse()` never calls `checksum()`; confirmed EGY absent from `countriesWithFormat`; ran full `npx jest` → 2156/2156; compared against Kuwait's `parse()`/`validate()` pattern)
- CI reproduced clean in an isolated worktree: Prettier ✓, ESLint ✓, `tsc` build ✓, Jest 2156/2156 ✓

## Strengths
- Thorough, transparent research doc (`docs/research/egypt-national-id.md`) — explicitly discloses the empirical-verification tradeoff, cites sources consulted, and even open-calls for a stronger citation. This is exactly the kind of documentation that makes the P1 #4 judgment call reviewable at all.
- Correct core logic: checksum math, leap-year/date validation, all 27 governorate codes + `88`, century resolution, gender parity, and registry wiring (`registerAll.ts`, `index.ts`, migration fixtures) all verified correct.
- Clean history of iterating on reviewer feedback (e.g. `fix(#54): correct Luhn parity ... (Codex review)`).

## Next Steps
- [ ] Fix P1 #1: make `parse()` verify the checksum (return `null` on mismatch), matching the Kuwait convention
- [ ] Fix P1 #2: add EGY to `countriesWithFormat` in `getCountryIdFormat-migration.test.ts`
- [ ] Fix P1 #3: bump README test count 2105 → 2156
- [ ] Address P1 #4: add a README caveat and/or reconsider default-enforced checksum, per the Bahrain precedent
- [ ] Consider P2 suggestions
- [ ] Re-request review

**Comment by ramezhanna** [comment]
Addressed in `f97dc21` + `fed3948`. Point-by-point:

**P1 #1 — `parse()` didn't verify the checksum.** Fixed. `parse()` now computes `checksum(normalized)` and returns `null` on mismatch, so it agrees with `validate()` on every input. Matches the Kuwait `parse()`/`validate()` convention. Added a regression test asserting `parse('29001010100010')` is `null` and consistent with `validate()`.

**P1 #2 — EGY missing from `countriesWithFormat`.** Fixed. Added `{ code: 'EGY', format: 'CYYMMDDGGSSSSV' }`, so CI now pins the display format.

**P1 #3 — stale README test count.** Fixed. Bumped to **2158** (full `npx jest` on the branch after these changes: 2158 passed — the +2 over your 2156 are the new format-fixture row and the parse-checksum regression test).

**P1 #4 — checksum enforced-by-default, undisclosed in README.** Kept enforcement (deliberate design choice) and added the missing disclosure: the README Egypt line now points to the research note documenting that the check digit is empirically derived, not an official spec. The research doc was also updated to redact the real-ID references.

**P2 #1 — `parse()` docstring "(NOT the opt-in checksum)".** Fixed — docstring now describes the enforced full contract.

**P2 #2 — "All 80 registered countries" comment.** Fixed → 81.

**P2 #3 — commits missing `(#54)` tag.** Noted; those are already-published commits so I haven't rewritten history, but new commits use the proper tag.

**P2 #4 — `nzl/irdNumber.ts` stale comment.** Out of scope for this PR (pre-existing, untouched), leaving as-is per your note.

CI green on the branch: Prettier &#10003;, ESLint &#10003;, `tsc` &#10003;, Jest 2158/2158 &#10003;. Re-requesting review.

File: src/countries/egy/nationalId.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  const year = resolveYear(century, yy);
+  if (year === null || !isValidDate(year, parseInt(mm, 10), parseInt(dd, 10))) {
+    return false;
+  }
+
+  if (!(gov in GOVERNORATES)) {
+    return false;
+  }
+
+  const expected = checksum(normalized);
```
Comment (AngusHsu-cubewise) [inline src/countries/egy/nationalId.ts]: **P1 — this rejects IDs using a checksum formula that has no authoritative or reproducible validation corpus.** The accompanying research document concedes no government specification was found; its five claimed matches are private, and every repository test vector is generated by this exact formula. Please do not make this a validation gate (`hasChecksum: true`) until it can be matched to Python upstream or an authoritative, independently reproducible source. Otherwise genuine IDs that use the undisclosed rule—including every payload this formula maps to `10`—will be rejected.

**Review by AngusHsu-cubewise — CHANGES_REQUESTED** [review]
## Changes requested

### P1 — Do not enforce the unverified checksum

This PR changes `validate()`/`parse()` from structural validation to rejecting every ID whose final digit fails a privately inferred mod-11 rule. The research record explicitly says no government specification was found and the only five purported matching IDs cannot be shared; all checked-in vectors are generated by this same implementation, so they only demonstrate self-consistency. An independent Egyptian-ID provider likewise documents that full check-digit validation requires the government’s proprietary algorithm and recommends structural validation only ([source](https://api.signme.it/blog/egyptian-id-fields-explained#how-do-i-validate-a-national-id-number-format-before-calling-the-api)).

This can produce false negatives for genuine IDs, including the code path that asserts an input payload can never be valid when the formula yields `10`. It also violates the repository’s source-of-truth rule: the PR documents that Python `idnumbers` has no EGY implementation to establish the behavior against.

Please revert to format/date/governorate validation with `hasChecksum: false` (and no checksum-gated `parse()`), as originally described in the PR, until the Python upstream or an authoritative reproducible specification/test corpus establishes the algorithm. Then add the verified checksum in both projects with independent vectors.

### P2 — Research record is now inconsistent with the implementation

`docs/research/egypt-national-id.md` still states twice that `parse()` does **not** gate on the checksum (lines 41–42 and 100–102), although the latest code does. Update it if checksum enforcement remains after the P1 is resolved.

I reran the current head in a clean worktree: Jest **2158/2158**, ESLint, TypeScript build, and Prettier all pass.

File: src/countries/egy/nationalId.ts (hunk 1 above)
Comment (ramezhanna) [reply src/countries/egy/nationalId.ts]: comment addresed

**Review by AngusHsu — CHANGES_REQUESTED** [review]
## Deep Review — Egypt National ID (EGY)

Reviewed at `c846d75`. Ran the four CI gates locally against the PR head: **Prettier ✅ · ESLint ✅ · tsc ✅ · Jest ✅ 2189/2189**.

First, credit where it's due: `docs/research/egypt-national-id.md` is an unusually rigorous piece of work. Identifying that the 5/5 checksum fit was underdetermined — that ~10⁸ weight vectors satisfy the same corpus, and that positions 8 and 10 are multiplied by zero across every sample — is exactly the right analysis, and reaching "do not implement a checksum" rather than shipping a plausible guess is the correct call. The single blocking issue below is that the code doesn't yet fully carry out that decision.

### Scores
- Security: 9/10
- Performance: 10/10
- Code Quality: 7/10
- Test Coverage: 8/10

### P0 — Critical
None.

### P1 — Important

**Egypt is the only country in the library that declares `checksum: false` yet returns a non-null check digit** — `src/countries/egy/nationalId.ts:134`, `:234`

The research doc concludes "Do not implement a checksum," and `aff8820` correctly stopped *enforcing* it. But the implementation is still exported and still reachable, so the decision is only half-applied.

To be clear about what is *not* the problem: exporting a `checksum` function alongside `hasChecksum: false` is fine and matches existing convention — `MYS`, `NGA`, `NZL`, `PAK` and `VNM` all do it. The convention those five follow is that the function **returns `null`**. `src/countries/nga/nationalId.ts:95` states the contract explicitly: the null return *"mirrors `checksum: false` in METADATA."*

Egypt is the sole exception across all 81 registered countries:

```
EGY: METADATA.checksum=false | checksum(29001010100017) -> 7      <-- unique
MYS: METADATA.checksum=false | checksum(800101011234)   -> null
NGA: METADATA.checksum=false | checksum(12345678901)    -> null
NZL: METADATA.checksum=false | checksum(AB123456)       -> null
PAK: METADATA.checksum=false | checksum(1234567890123)  -> null
VNM: METADATA.checksum=false | checksum(001099012345)   -> null
```

`createValidator()` in `src/registry/adapters.ts` wires any module-level `checksum` into the registry, so this reaches consumers as first-class API. The module docstring's guardrail ("Nothing in this module calls it") holds inside the module but not at the package boundary:

```
registry.getFormat('EGY').checksumAlgorithm  -> "None (check digit algorithm not publicly documented...)"
registry.get('EGY').METADATA.checksum        -> false
registry.get('EGY').checksum('29001010100017') -> 7
```

The concrete failure mode — a consumer writing the natural guard against this API:

```ts
const c = validator.checksum(id);
if (c !== null && c !== Number(id.slice(-1))) reject(id);
```

This behaves correctly for all five other `checksum: false` countries (the `null` short-circuits) and silently misfires **only** on Egypt — rejecting genuine IDs for every governorate outside `01`–`04` at a rate approaching 10/11. That is precisely the harm the research doc set out to prevent, and it's reachable today via the public registry. The doc also argues the shipped reduction `(11 - sum % 11) % 11` is itself likely wrong (emits an impossible `10` on ~9.2% of payloads; disagrees with both public implementations on ~18.3%), which makes a returned value worse than no value.

**Suggested fix** — match the five existing precedents:
```ts
export function checksum(_idNumber: string): number | null {
  return null; // No documented algorithm — see docs/research/egypt-national-id.md
}
```
Or drop `checksum` from the `NationalID` object entirely so the adapter can't publish it. Either way, `CHECKSUM_WEIGHTS` and the mod-11 body should go — the research record preserves the hypothesis far better than dead code does.

### P2 — Suggestions
- **"format-only" understates what `validate()` does** — `README.md:286`, `nationalId.ts:86` (`checksumAlgorithm`), and `issue-54-egy-nationalId.test.ts:36`. Validation also enforces a real calendar date and a known governorate. Suggest "format + semantic; no check-digit validation."
- **`GOVERNORATES: Record<string, string>`** (`:42`) overstates valid keys; `as const` plus a `code is GovernorateCode` guard would preserve the invariant at the type level.
- **Barrel export untested** — `export * as EGY from './countries/egy'` has no direct coverage; registry tests would still pass if it broke or were renamed.
- **`normalize()` coerces non-strings** (`:97`) via `String(idNumber)`; most validators guard with `typeof idNumber !== 'string'` first.
- **Migration-parity suites** (`parseIdInfo-migration.test.ts`, `validateNationalId-migration.test.ts`) state they verify old-vs-new switch parity; EGY has no switch-based predecessor. Either move the coverage to the Egypt suite or widen the suite comments.
- **Doc off-by-one** — "23 of the 28 governorate codes" mixes 23 (governorates excluding `88`) with 28 (codes including it). 24 of 28 codes are affected.

### Considered and dismissed
Flagged during review, checked, and found not to apply — noting them so they don't resurface:
- **"Future birth dates validate"** (`39912310100000` → `age: -74`). Factually true, but library-wide rather than an Egypt defect: all 14 countries using the shared `calculateAge` behave identically and none guard it. Belongs in its own issue.
- **"Doesn't match the Python source of truth."** Python has no Egypt module; the doc's RUS precedent covers non-upstream countries.
- **"81st registry key breaks the 80-country invariant."** That invariant guards against *secondary* ID types entering the registry; Egypt is a legitimate primary country, and the hard assertion is correctly updated to 81.
- **`pattern`/`isParsable`/`hasChecksum` vs `IdMetadata`.** Supported deliberately via `FunctionBasedMetadata` + `adaptMetadata()`.

### Housekeeping
The PR description is now stale and describes an API that no longer exists: it advertises an opt-in Luhn check via `validate(id, { strictChecksum: true })`, which `aff8820` removed. The test/count figures also drifted (says 36 cases / 2141 passing; the branch has 31 Egypt tests / 2189 total). Worth refreshing before merge so the merge commit doesn't record behavior that isn't shipped.

### Verdict: REQUEST CHANGES — one P1

The blocker is a one-line change that the PR's own research already argues for, with five in-repo precedents to copy. Everything else is optional polish. Nice work on the investigation.

**Comment by ramezhanna** [comment]
> eb184de

Fixed P1 in this comment [eb184de](https://github.com/identique/idnumbers-npm/pull/114/commits/eb184dea1ef61aca9e0343858340d0b02c045ad5). The P2 suggestion has been temporarily removed, as there is no current requirement to address it.

**Review by AngusHsu — CHANGES_REQUESTED** [review]
## Deep Review — Egypt National ID (EGY)

Reviewed at `f5cd715` (10-stream deep review).

### First, credit where it's due

@ramezhanna — this PR is on its fourth review round, and it's worth stating plainly that most of that churn was **not** your fault. The first two rounds pointed in opposite directions:

- **Round 1** (07-06) flagged that `parse()` didn't verify the checksum and so disagreed with `validate()` — which led to `e7e1920` and `f97dc21` enforcing it in both.
- **Round 2** (07-15) then asked that the checksum not be enforced *at all* — discarding that work.

Four commits lost to contradictory guidance from two different reviewers. That's a process failure on our side, not an implementation failure on yours.

`docs/research/egypt-national-id.md` remains the strongest artifact in this PR. Recognising that the 5/5 checksum fit was *underdetermined* — that ~10⁸ weight vectors satisfy the same corpus, and that positions 8 and 10 are multiplied by zero in every available sample — and then concluding **"do not implement a checksum"** rather than shipping a plausible guess is the correct call, and a harder one than shipping it would have been. Thanks also to @AngusHsu-cubewise for the round-2 catch that redirected this toward structural validation.

### Where things actually stand

The blocker has shrunk every round. This is the last mile, not a stuck PR:

| Round | Blocker |
|---|---|
| 2 | Enforces a guessed algorithm → rejects genuine IDs |
| 3 | No longer enforces, but still *returns* a guessed non-null value |
| **4 (this)** | Returns `null` correctly — but the **documentation** still describes the computation |

### P0 — Critical
None.

### P1 — The one blocker

**`checksum()` is documented as computing a mod-11 digit but unconditionally returns `null`** — `src/countries/egy/nationalId.ts:132`

Round 3 asked for two changes. `eb184de` applied one of them:

- ✅ `checksum()` returns `null` — matching the `MYS`/`NGA`/`NZL`/`PAK`/`VNM` precedent
- ❌ *"`CHECKSUM_WEIGHTS` and the mod-11 body should go"* — they were **commented out, not deleted**

That single choice is what produced this finding. Because the dead algorithm still sits in the file as comment text, three separate doc sites still read as though it runs:

1. **File docblock** — *"A mod-11 hypothesis is exposed via `checksum()` for research purposes"* — it is not exposed; it returns `null`.
2. **Function JSDoc:112** — *"Compute a mod-11 check digit over the first 13 digits"*, and *"Treat the weights above as one arbitrary pick"* — pointing at weights that are now commented out.
3. **`NationalID.checksum` label** — *"UNVERIFIED research helper"* — implies it computes something.

A consumer reading any of the three and calling `checksum()` gets `null` for every input. Round 3 anticipated exactly this: *"the research record preserves the hypothesis far better than dead code does."*

**Fix** — delete rather than comment:

```ts
/**
 * Egypt publishes no official check-digit algorithm and no independently
 * reproducible specification could be located, so no checksum is computed.
 * Always returns null, mirroring `checksum: false` in METADATA.
 * The mod-11 hypothesis and why it was rejected: docs/research/egypt-national-id.md
 */
export function checksum(_idNumber: string): number | null {
  return null;
}
```

Then remove the commented-out `CHECKSUM_WEIGHTS` block and update the file docblock and the `NationalID.checksum` label to match.

### P2 — Carried forward from round 3, still open

These five have appeared in consecutive reviews without being addressed. Worth clearing in the same commit so they stop resurfacing:

- **`"format-only"` understates `validate()`** — `README.md:286`, `nationalId.ts:86` (`checksumAlgorithm`), `issue-54-egy-nationalId.test.ts:36`. Validation also enforces a real calendar date and a known governorate. Suggest *"format + semantic; no check-digit validation."*
- **`GOVERNORATES: Record<string, string>`** (`:42`) overstates valid keys — `as const` plus a `code is GovernorateCode` guard preserves the invariant at type level.
- **Barrel export untested** — `export * as EGY from './countries/egy'` has no direct coverage; registry tests would still pass if it broke or were renamed.
- **`normalize()` coerces non-strings** (`:97`) via `String(idNumber)`; most validators guard with `typeof idNumber !== 'string'` first.
- **Migration-parity suite placement** — `country-specific-validation.test.ts:401` is scoped to Python `idnumbers` parity, but Python has no `egy` module. Move the cases to `issue-54-egy-nationalId.test.ts`, or widen the suite comment.

Also still open from round 3: the **PR description is stale** — it advertises `validate(id, { strictChecksum: true })`, an API that `aff8820` removed, and the test counts have drifted. Worth refreshing so the merge commit doesn't record behavior that isn't shipped.

### Settled — please do not re-raise

Round 3 examined and dismissed these. Restating so future reviews stop spending findings on them:

- **"81st registry key breaks the 80-country invariant."** The invariant guards against *secondary* ID types entering the registry. Egypt is a legitimate primary country and the hard assertion is correctly updated to 81. **Not a defect.**
- **"Doesn't match the Python source of truth."** Python has no Egypt module; the RUS precedent covers non-upstream countries.
- **"Future birth dates validate."** True but library-wide — all 14 countries using the shared `calculateAge` behave identically. Belongs in its own issue.
- **`pattern` / `isParsable` / `hasChecksum` vs `IdMetadata`.** Supported deliberately via `FunctionBasedMetadata` + `adaptMetadata()`.

### Path to green

One commit closes this out:

1. Delete `CHECKSUM_WEIGHTS` and the commented-out mod-11 body
2. Rewrite the three doc sites to state `checksum()` is intentionally unimplemented
3. Sweep the five carried-forward P2s
4. Refresh the stale PR description

### Verdict: REQUEST CHANGES — one P1

The blocker is a deletion plus three comment edits, and the PR's own research already argues for it. Genuinely good investigative work here — the hard part was decided correctly several rounds ago; what remains is making the prose agree with the code.

**Comment by ramezhanna** [comment]
P1 — CHECKSUM_WEIGHTS and the mod-11 body are deleted rather than commented; checksum() is now an unconditional return null. The file docblock, function JSDoc and NationalID.checksum label all state it is intentionally unimplemented. The now-unused validateRegexp import is dropped. The hypothesis and the reasoning that rejected it stay in the research doc.
One deviation: the suggested _idNumber still trips @typescript-eslint/no-unused-vars — the repo's eslint config sets it to error with no argsIgnorePattern — so I used the eslint-disable-next-line form already used by nga/nationalId.ts:102.

P2 — all five: "format-only" wording corrected in README, checksumAlgorithm and the test comment; GOVERNORATES is as const with an isGovernorateCode guard; barrel export now has a direct test; normalize() rejects non-string input instead of coercing; EGY cases moved out of the Python-parity suite. PR description refreshed.
Also added a test asserting checksum() returns null — coverage showed nothing exercised it.
2193 tests passing; build, lint and format:check clean.

**Review by AngusHsu — DISMISSED** [review]
## Deep Review — Egypt National ID (EGY)

Multi-stream automated review (10 parallel streams: general, silent-failure, test-coverage, type-design, comment-accuracy, guideline-compliance, bug-scan, history-context, prior-PR-patterns, comment-compliance). All findings were re-verified against head `7a386be` before reporting.

### Scores
- Security: 10/10
- Performance: 10/10
- Code Quality: 9/10
- Test Coverage: 9/10

### Verification run at head
| Check | Result |
|---|---|
| Prettier | ✅ All files formatted |
| TypeScript (`tsc --noEmit`) | ✅ exit 0 |
| Jest | ✅ 29 suites, **2193 tests** passed |
| ESLint (PR-touched files) | ✅ 0 errors — the one `no-explicit-any` warning in `src/index.ts` is pre-existing, not introduced here |

### P0 — Critical
None.

### P1 — Important
None.

### P2 — Suggestions

**1. `README.md:477` — stale test count.** Reads "2189 tests"; the actual Jest total at this head is **2193**. Repo convention is to bump this line to the real total whenever tests are added.

**2. `src/__tests__/validateNationalId-migration.test.ts:416` and `src/__tests__/parseIdInfo-migration.test.ts:358` — parity-suite intent mismatch.** Both suites are documented as verifying that the registry-based implementation produces *identical results to the previous switch-based implementation*. EGY is new and has no switch-based predecessor, so its inclusion doesn't test migration parity. Either keep EGY coverage in `issue-54-egy-nationalId.test.ts`, or update the suite comments to state they now also carry post-migration registry coverage.

**3. `src/countries/egy/nationalId.ts:76` — `isGovernorateCode` type guard (nit).** `code in GOVERNORATES` walks the prototype chain, so it's technically unsound as a narrowing predicate. Not reachable in practice — the regex constrains `gov` to `\d{2}`, and no prototype key matches that — so this is purely a strictness nit. `Object.prototype.hasOwnProperty.call(GOVERNORATES, code)` would close it.

### Informational — no action required

- **Registry now holds 81 primary keys.** The count assertion in `parseIdInfo-migration.test.ts` was updated in lockstep, and `README.md` country counts and the "Africa (4)" section header were all updated consistently. Correct and complete.
- **No upstream Python counterpart.** The Python `idnumbers` library has no `egy` module, so this implementation has no upstream port to mirror against. Flagging only so maintainers can consciously ratify expanding Node coverage ahead of Python, rather than having it happen implicitly.
- **Future birth dates are accepted** (century digit `3` permits dates through 2099). Consistent with the existing codebase — only 2 of 22 date-parsing country modules guard against future dates — so this is a repo-wide convention question, not a defect in this PR.

### Resolved from earlier review rounds
Two concerns raised against earlier commits on this branch no longer apply at head, and are explicitly **not** findings:

- *"`hasChecksum: false` while `checksum` is still exported"* — the current state is coherent and deliberate: `checksum()` returns `null`, its JSDoc explains the function is retained for signature parity with other country modules, `checksumAlgorithm` documents the gap in user-visible metadata, and a test covers the null return.
- *"README Egypt row omits a checksum-limitation caveat"* — BHR, the format-only precedent cited in the research doc, carries no caveat in its README row either. Consistent with existing convention.

### Verdict: PASS ✅

No P0 or P1 issues survive verification. This is careful work — the decision to withdraw the unverified mod-11 check digit rather than ship a guessed algorithm is the right call and is honestly documented in `docs/research/egypt-national-id.md`, including why the hypothesis was rejected. Error paths are explicit throughout, and `normalize()` rejecting non-string input instead of coercing it is a nice touch that prevents a numeric literal slipping through as a string.

Only P2 #1 (the README test count) is worth fixing before merge; the other two are optional.

**Comment by ramezhanna** [comment]
Thanks for the thorough review, @AngusHsu! Addressed all three P2 suggestions in d933d73:

1. **`README.md` test count** — bumped 2189 → 2193 (verified against the actual Jest total).
2. **Parity-suite intent mismatch** — kept the EGY coverage and updated the header comments in both `parseIdInfo-migration.test.ts` and `validateNationalId-migration.test.ts` to state they also carry post-migration registry coverage for countries (like EGY) that never had a switch-based predecessor.
3. **`isGovernorateCode` type guard** — switched `code in GOVERNORATES` to `Object.prototype.hasOwnProperty.call(GOVERNORATES, code)` to close the prototype-chain soundness gap.

All 2193 tests pass, tsc clean, prettier clean.
